File: scripts/health-check.py
#!/usr/bin/env python3
"""
Health check script to verify the backend is running before launching the frontend.
"""

import sys
import time
import subprocess
import requests
import os
import threading
import signal
import math

BACKEND_URL = os.environ.get('BACKEND_URL', 'http://localhost:8000')
HEALTH_ENDPOINT = f"{BACKEND_URL}/health"
MAX_RETRIES = 30
INITIAL_RETRY_INTERVAL = 2  # seconds
TIMEOUT_PER_CHECK = 5  # seconds

def check_health():
    """Check if the backend health endpoint is responding."""
    retry_count = 0
    current_interval = INITIAL_RETRY_INTERVAL

    while retry_count < MAX_RETRIES:
        try:
            print(f"[INFO] Attempting health check ({retry_count + 1}/{MAX_RETRIES})...")

            response = requests.get(HEALTH_ENDPOINT, timeout=TIMEOUT_PER_CHECK)

            if response.status_code == 200:
                print("[SUCCESS] Backend health check passed!")
                print(f"[INFO] Response: {response.json()}")
                return True
            elif response.status_code == 404:
                print(f"[ERROR] Health endpoint not found: {HEALTH_ENDPOINT}")
            else:
                print(f"[ERROR] Backend health check failed with status: {response.status_code}")

        except requests.exceptions.ConnectionError:
            print(f"[WARN] Backend not reachable yet: Connection refused")
        except requests.exceptions.Timeout:
            print(f"[WARN] Backend health check timed out after {TIMEOUT_PER_CHECK}s")
        except requests.exceptions.RequestException as e:
            print(f"[WARN] Request error during health check: {str(e)}")
        except Exception as e:
            print(f"[WARN] Unexpected error during health check: {str(e)}")

        retry_count += 1
        if retry_count < MAX_RETRIES:
            # Exponential backoff with a maximum cap
            current_interval = min(current_interval * 1.1, 10)  # Cap at 10 seconds
            print(f"[INFO] Waiting {current_interval:.1f}s before retry {retry_count + 1}/{MAX_RETRIES}...")
            time.sleep(current_interval)

    print("[ERROR] Backend health check failed after maximum retries")
    return False

def main():
    """Main function to check backend health and execute command if healthy."""
    print("[INFO] Starting health check process...")
    print(f"[INFO] Health endpoint: {HEALTH_ENDPOINT}")
    print(f"[INFO] Maximum retries: {MAX_RETRIES}")
    print(f"[INFO] Initial retry interval: {INITIAL_RETRY_INTERVAL}s")
    print(f"[INFO] Timeout per check: {TIMEOUT_PER_CHECK}s")

    # Check if the health endpoint is available
    if check_health():
        print("[SUCCESS] Backend is healthy and ready!")

        # Execute the original command that was passed to this script
        if len(sys.argv) > 1:
            command = sys.argv[1:]
            print(f"[INFO] Executing command: {' '.join(command)}")

            try:
                # Execute the command and wait for it to complete
                result = subprocess.run(command, check=True)
                print(f"[INFO] Command completed successfully with return code: {result.returncode}")
                sys.exit(result.returncode)
            except subprocess.CalledProcessError as e:
                print(f"[ERROR] Command failed with return code: {e.returncode}")
                sys.exit(e.returncode)
            except KeyboardInterrupt:
                print("[INFO] Command interrupted by user")
                sys.exit(130)  # Standard exit code for SIGINT
        else:
            print("[INFO] Backend is healthy, no command to execute")
            return 0
    else:
        print("[ERROR] Backend health check failed, cannot proceed.")
        print("[ERROR] Exiting with code 1")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: scripts/health-check.js
#!/usr/bin/env node

const http = require('http');
const https = require('https');
const { execSync, spawn } = require('child_process');

const BACKEND_URL = process.env.BACKEND_URL || 'http://localhost:8000';
const HEALTH_ENDPOINT = `${BACKEND_URL}/health`;
const MAX_RETRIES = 30;
const INITIAL_RETRY_INTERVAL = 2000; // milliseconds
const TIMEOUT_PER_CHECK = 5000; // milliseconds

function checkHealth(retryCount = 0, currentInterval = INITIAL_RETRY_INTERVAL) {
  return new Promise((resolve, reject) => {
    const url = new URL(HEALTH_ENDPOINT);

    // Choose http or https based on protocol
    const client = url.protocol === 'https:' ? https : http;

    const options = {
      hostname: url.hostname,
      port: url.port,
      path: url.pathname,
      method: 'GET',
      timeout: TIMEOUT_PER_CHECK
    };

    const req = client.request(options, (res) => {
      if (res.statusCode === 200) {
        console.log('✅ Backend health check passed!');
        // Try to read response data
        let data = '';
        res.on('data', chunk => data += chunk);
        res.on('end', () => {
          try {
            const jsonData = JSON.parse(data);
            console.log('📍 Response:', jsonData);
          } catch (e) {
            console.log('📍 Response text:', data.substring(0, 200) + (data.length > 200 ? '...' : ''));
          }
          resolve(true);
        });
      } else if (res.statusCode === 404) {
        console.log(`❌ Health endpoint not found: ${HEALTH_ENDPOINT}`);
        if (retryCount < MAX_RETRIES) {
          const nextInterval = Math.min(currentInterval * 1.1, 10000); // Cap at 10 seconds
          console.log(`⏳ Retrying in ${(nextInterval / 1000).toFixed(1)}s... (${retryCount + 1}/${MAX_RETRIES})`);
          setTimeout(() => checkHealth(retryCount + 1, nextInterval).then(resolve).catch(reject), nextInterval);
        } else {
          reject(new Error('Backend health check failed after maximum retries'));
        }
      } else {
        console.log(`❌ Backend health check failed with status: ${res.statusCode}`);
        if (retryCount < MAX_RETRIES) {
          const nextInterval = Math.min(currentInterval * 1.1, 10000); // Cap at 10 seconds
          console.log(`⏳ Retrying in ${(nextInterval / 1000).toFixed(1)}s... (${retryCount + 1}/${MAX_RETRIES})`);
          setTimeout(() => checkHealth(retryCount + 1, nextInterval).then(resolve).catch(reject), nextInterval);
        } else {
          reject(new Error('Backend health check failed after maximum retries'));
        }
      }
    });

    req.on('error', (err) => {
      console.log(`⚠️  Backend not reachable yet: ${err.message}`);
      if (retryCount < MAX_RETRIES) {
        const nextInterval = Math.min(currentInterval * 1.1, 10000); // Cap at 10 seconds
        console.log(`⏳ Retrying in ${(nextInterval / 1000).toFixed(1)}s... (${retryCount + 1}/${MAX_RETRIES})`);
        setTimeout(() => checkHealth(retryCount + 1, nextInterval).then(resolve).catch(reject), nextInterval);
      } else {
        reject(new Error('Backend health check failed after maximum retries'));
      }
    });

    req.on('timeout', () => {
      console.log('⏰ Backend health check timed out');
      req.destroy();
      if (retryCount < MAX_RETRIES) {
        const nextInterval = Math.min(currentInterval * 1.1, 10000); // Cap at 10 seconds
        console.log(`⏳ Retrying in ${(nextInterval / 1000).toFixed(1)}s... (${retryCount + 1}/${MAX_RETRIES})`);
        setTimeout(() => checkHealth(retryCount + 1, nextInterval).then(resolve).catch(reject), nextInterval);
      } else {
        reject(new Error('Backend health check failed after maximum retries'));
      }
    });

    req.end();
  });
}

async function main() {
  console.log('🔍 Starting health check process...');
  console.log(`📍 Health endpoint: ${HEALTH_ENDPOINT}`);
  console.log(`📍 Maximum retries: ${MAX_RETRIES}`);
  console.log(`📍 Initial retry interval: ${(INITIAL_RETRY_INTERVAL / 1000)}s`);
  console.log(`📍 Timeout per check: ${(TIMEOUT_PER_CHECK / 1000)}s`);

  try {
    await checkHealth();
    console.log('🎉 Backend is healthy and ready!');

    // Execute the original command that was passed to this script
    if (process.argv.length > 2) {
      const command = process.argv.slice(2).join(' ');
      console.log(`🚀 Executing command: ${command}`);

      try {
        const child = execSync(command, { stdio: 'inherit' });
        console.log('✅ Command completed successfully');
        process.exit(0);
      } catch (error) {
        console.error('💥 Command failed:', error.message);
        process.exit(error.status || 1);
      }
    } else {
      console.log('ℹ️  Backend is healthy, no command to execute');
    }
  } catch (error) {
    console.error('💥 Backend health check failed:', error.message);
    console.error('❌ Exiting with code 1');
    process.exit(1);
  }
}

main().catch(console.error);